### DISCUSSION TITLE
Is it possible to evaluate a variable at a given point?

### DISCUSSION THREAD

**Discussion by Adrien-a0** [MPh-py/MPh#145] [General]
I'm trying to evaluate the value of one variable at the minima of another, which I obtain by fitting the variable using scipy. I would like to be able to evaluate the point directly at this position using MpH, however I don't know if I can. My current work around is to export data at the point by defining the point using global variables in COMSOL. This, however, requires the solution to be computed twice, due to the use of global variables.

Alternatively, if it is possible to assign values variables that are only defined in the results interface I could also obtain this value using only one computation by using my current workaround.

**Comment by john-hen** [comment]
>My current work around is to export data at the point by defining the point using global variables in COMSOL.

If I understand correctly (and that's difficult without a concrete example), you already know how to evaluate a (field) variable at a given point. I'm assuming when you say "export" you mean "evaluate", and by "global variable" you mean "global parameter". (Otherwise that's what you should look into.)

But then you don't have to compute the solution twice. Comsol let's you update the solution when parameters change. This feature arguably exist precisely for this use case, where the parameter values are only needed in post-processing.

**Reply by Adrien-a0** [reply]
Hey, thanks for your reply. No, I mean literally export it to a .txt file. I have an export named 'data' that exports the data at one position (x_0,y_0,0) where x_0 and y_0 are global parameters. I use the lines of code:
```
model.parameter('x_0', f'{x_0}[m])
model.parameter('y_0', f'{y_0}[m])
model.solve('Std1')
model.export('data', file_name)

```
where I have already used the line of code 
`model.solve('Std1')`
to solve the model and fit for the minimum, but I have  to repeat the step in order to export the data since x_0 and y_0 are global parameters. 

In COMSOL it is possible to define x_0 and y_0 as local parameters within the results interface, however I don't know if it is possible to vary these variables using mph. Ideally when using the model.evaluate() I would be able to specify given coordinates though.

**Reply by Adrien-a0** [reply]
Just to clarify, I export to .txt file and then load the .txt file within python as I'm interested in processing the data using python, but as far as I can tell you can't evaluate a model at a specific point using mph.

**Reply by john-hen** [reply]
You have [access to the full Comsol API](https://mph.readthedocs.io/en/1.2/demonstrations.html#access-the-full-comsol-api) via MPh, so all of these things should be possible. Almost certainly *are* possible. It's just that MPh does not expose any "high-level" functions for this specific use case. But you can dig deeper, like in the Comsol documentation, or by inspecting the Java code of your model.

**Reply by Adrien-a0** [reply]
Hey I had a quick look over this stuff but I think it's a bit above my ability coding wise. I wonder if you know how I could update the solution of a study (named "Std1") in MPh, rather than recomputing? This method works far more quickly within the comsol GUI, so it would be hugely appreciated if you know a line of code that will do this!

**Reply by john-hen** [reply]
Looks like you'll have to call the Java method `updateSolution()` on the (Java) solution object.

For example, using the demo model from the tutorial (`capacitor.mph`), to update the "electrostatic solution", it's either
```python
model.java.sol('sol1').updateSolution()
```
or, using the "node syntax" introduced by MPh:
```python
(model/'solutions'/'electrostatic solution').java.updateSolution()
```
In both cases we access the Comsol API directly via the respective `.java` attribute.

You'll have to find out the solution name ("electrostatic solution" above) or tag ("sol1" above), which isn't the same as the study name or tag. Easiest is to hard-code it. To find out, if you don't know already, open your model in the Comsol GUI, do "update solution", then "save as" a `.java` file and look for the solution tag at the very end of the file. Or run `mph.tree(model)` in an interactive Python session to find the solution's name. Obviously, it would also be possible to have your program do the look-up, but that takes a little more effort to code.

**Reply by Adrien-a0** [reply]
Hey,

Thank you so much this works perfectly and was incredibly useful in understanding how to access the full api.
